README.md

```diff
@@ -514,29 +533,13 @@ Default UI options
     diagnostic = "🐞",
     incoming = " ",
     outgoing = " ",
-    colors = {
-      -- Normal background color for floating window
-      normal_bg = "#1d1536",
-      -- Title background color
-      title_bg = "#afd700",
-      red = "#e95678",
-      magenta = "#b33076",
-      orange = "#FF8700",
-      yellow = "#f7bb3b",
-      green = "#afd700",
-      cyan = "#36d0e0",
-      blue = "#61afef",
-      purple = "#CBA6F7",
-      white = "#d1d4cf",
-      black = "#1c1c19",
-    },
+    hover = ' ',
     kind = {},
```

Comment:
So currently catppuccin's lspsaga kind integration looks like this: https://github.com/catppuccin/nvim/blob/main/lua/catppuccin/groups/integrations/lsp_saga.lua#L21

Did `kind` become a single array with only kinds, I can't find document about this.

Reply:
it just remove colors in config. these colors still used for kind and default highlight groups.

Reply:
Lspsaga's winbar always gave white background after `ColorScheme` event, did this PR fix the issue?

Reply:
No it doesn't (Tried out this PR)

Before changing colorscheme

![image](https://user-images.githubusercontent.com/56817415/215261076-6d2d2458-4ae4-4778-9aea-de7b44faf888.png)

After changing colorscheme 

![image](https://user-images.githubusercontent.com/56817415/215261034-5c21ba37-00f6-4f56-833e-d8a4fc34e0a7.png)

Reply:
when you load new colorscheme it will have `hi clear` command to clear all highlight group . I don't think problem should be here ?  unless you defined all winbar highlight group in all of your colorschems or  callback like the comment 
do 

```lua
vim.api.nvim_create_autocmd("ColorScheme", {
  callback = function() require("lspsaga.highlight").init_highlight() end,
})
```

Reply:
Other plugins have these autocmd on by default.

Reply:
![image](https://user-images.githubusercontent.com/56817415/215261353-d72668f5-31bd-4a28-a1cc-3e3d011ae343.png)

Reply:
I would opt for removing the colors in `kind` config all together since they do not work if I change to other colorschemes (It keeps the old color table)

Users who want to change highlights can do so through `LspsagaWinbarProperty` if they wish

Otherwise you can support a function like so

```lua
kind = function()
return
end
```

and kind is auto-reloaded on `ColorScheme` event which is precisely how bufferline.nvim currently works

Reply:
I don't think this should plugin do.  support by other way.

Reply:
Pardon me, did you mean "I don't think this plugin should support the other way."?

Reply:
Seems to work 👍🏻

Reply:
Files didn't get redraw...

Reply:
https://github.com/glepnir/lspsaga.nvim/blob/35815b8aa8d733741130389d91d929ffe9c05223/lua/lspsaga/symbolwinbar.lua#L59

Reply:
which files ? it just redraw winbar highlight. files colorscheme not the lspsaga control

Reply:
Through `LspSagaWinbarFileIcon` highlight

Reply:
I will be more clear

Reply:
when you move the cursor it not work ?

Reply:
when you move the cursor it not work ?

Reply:
it taka color from devicons if there is no value it no work.

Reply:
Here is how it should look with catppuccin-latte
![image](https://user-images.githubusercontent.com/56817415/215262088-03c430b0-fdc9-47d3-854f-3d84fe07428d.png)

Here is how it look after calling `:colorscheme catppuccin-latte` (Previous colorscheme is catppuccin-mocha)

![image](https://user-images.githubusercontent.com/56817415/215262151-945f8c00-f42f-4288-bc6c-4f179ddee80d.png)

Reply:
Notice that `LspSagaWinbarFileIcon` stays the same after changing from dark to light mode

Reply:
when you move cursor it will restore i think

Reply:
So it's because `kind` is an array and not a function so it did not get reloaded after 

```lua
kind = require('catppuccin.groups.integrations.lsp_saga').custom_kind(),
```

Which is the reason why I suggest you support kind as a function so it can be reloaded after changing colorscheme

Reply:
nope because it have `default = true` attar.

Reply:
> when you move cursor it will restore i think

It does not because the user config table is fixed at startup, instead I suggested that you support function type for kind:

```lua
kind = function()
  require('catppuccin.groups.integrations.lsp_saga').custom_kind()
end
```

So that it can be reloaded after `ColorScheme` autocmd

Reply:
because it's a colorscheme  plugin.

Reply:
bug not here

Reply:
How else do I do it that the colors get reloaded?

Reply:
i missed  a place need a patch .

Reply:
I don't want to tell users that lspsaga's `kind` array is always fixed so it will be broken if you change colorscheme's flavour

Reply:
fixed.

Reply:
Since you are against the idea of setting ColorScheme events and I think these patches (nvim_hl_ns_fast) are just workarounds

I just wish I can handle these colors through highlight groups, not through some special arrays

Reply:
It still doesn't work (as a colorscheme plugin author)

```
kind = require('catppuccin.groups.integrations.lsp_saga').custom_kind()
```

Which is equals to

```
:lua=require('catppuccin.groups.integrations.lsp_saga').custom_kind()                            
{                                                                                                
  Array = { " ", "#89B4FA" },                                                                   
  Boolean = { " ", "#FAB387" },                                                                 
  Class = { " ", "#CBA6F7" },                                                                   
  Constant = { " ", "#89DCEB" },                                                                
  Constructor = { " ", "#89B4FA" },                                                             
  Enum = { "了", "#A6E3A1" },                                                                    
  EnumMember = { " ", "#A6E3A1" },                                                              
  Event = { " ", "#CBA6F7" },                                                                   
  Field = { " ", "#94E2D5" },                                                                   
  File = { " ", "#CDD6F4" },                                                                    
  Function = { " ", "#CBA6F7" },                                                                
  Interface = { " ", "#FAB387" },                                                               
  Key = { " ", "#F38BA8" },                                                                     
  Macro = { " ", "#F38BA8" },                                                                   
  Method = { " ", "#CBA6F7" },                                                                  
  Module = { " ", "#89B4FA" },                                                                  
  Namespace = { " ", "#FAB387" },                                                               
  Null = { " ", "#F38BA8" },                                                                    
  Number = { " ", "#A6E3A1" },                                                                  
  Object = { " ", "#FAB387" },                                                                  
  Operator = { " ", "#A6E3A1" },                                                                
  Package = { " ", "#CBA6F7" },                                                                 
  Parameter = { " ", "#89B4FA" },                                                               
  Property = { " ", "#89DCEB" },                                                                
  StaticMethod = { "ﴂ ", "#FAB387" },                                                            
  String = { " ", "#A6E3A1" },                                                                  
  Struct = { " ", "#CBA6F7" },                                                                  
  TypeAlias = { " ", "#A6E3A1" },                                                               
  TypeParameter = { " ", "#A6E3A1" },                                                           
  Variable = { " ", "#89B4FA" }                                                                 
}
```

Now you see where I am going, the colors are fixed forever, so if I want to switch to a light theme for example the colors remain the same

Reply:
you don't need both have `default` attar if your colorscheme defined it will use them . and your colorscheme override the lspsaga kind color. then  it will use these color even you changed color ? what do you want ? i think the redraw issue had fixed ?

Reply:
Aren't kind colors **supposed** to get overriden?

Reply:
hmm I don't think i got your point ? you mean when you change the colorscheme like dark to light the color also changed right ?

Reply:
It is supposed to change the `kind` array too because it has changed from dark to light

Reply:
But because the kind array is fixed at startup you can't really override it

Reply:
[Screencast from 2023-01-28 18-03-23.webm](https://user-images.githubusercontent.com/56817415/215263140-fb6c9627-932a-46d3-b2b6-15a64ddbf19b.webm)

Reply:
should be works now.

Reply:
Still need `kind` to support a function

Reply:
Ideally please remove the colors from `kind`, this is very problematic with monkeypatched code to redraw

If an user want to override the colors they will just do it through these highlights

`LspsagaWinbarProperty`

`LspsagaWinbarClass`

`LspsagaWinbarStruct`

Reply:
this should work and can be add into callback
```lua
package.loaded['lspsaga.lspkind'] = nil
require('lspsaga.kind').get_kind
```

Reply:
This will also helps removing lspsaga.nvim from special integration and the highlights will "Just works"

Reply:
> this should work and can be add into callback
> 
> ```lua
> package.loaded['lspsaga.lspkind'] = nil
> require('lspsaga.kind').get_kind
> ```

It doesn't matter because currently the only way for a colorscheme to set winbar highlights is via:

```lua
require('lspsaga').setup {
  ui = {
    kind = require('catppuccin.groups.integrations.lsp_saga').custom_kind(),
  }
}
```

As you can see the setup table is fixed FOREVER, doesn't matter if you reload the module or not, user config stayed the same

And as fair as I know this is the only way a colorscheme can integrate into lspsaga's winbar which is a little suboptimal.

Reply:
```lua
require('lspsaga').setup {
  kind = function()
    require(vim.g.colors_name).palette()
  end
}
```

But if you change the kind to a function however, you can reload it even after setup and it will reload accordingly

Reply:
In conclusion, [nvim-navic](https://github.com/SmiteshP/nvim-navic/#-customise)'s style works the best here.

Reply:
why i don't public it because I don't like these string in my memory .I only want them run once. all of my plugin will clean the memory after the command or shotcut use.  i like a clean and lightweight things.

Reply:
this way ignored the your feature dynamic changed.  so i need find a better way to support it .

Reply:
Clean and lightweight thing in this case would be to remove all the colors in kind config and let user set them through neovim's native highlights

I don't want to leave a note like `keep in mind that symbolwinbar won't be able to reload colors`

Reply:
so need more smart way to do it .

Reply:
So what's wrong with

```lua
vim.api.nvim_set_hl(0, "LspsagaWinbarFile",          {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarModule",        {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarNamespace",     {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarPackage",       {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarClass",         {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarMethod",        {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarProperty",      {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarField",         {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarConstructor",   {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarEnum",          {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarInterface",     {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarFunction",      {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarVariable",      {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarConstant",      {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarString",        {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarNumber",        {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarBoolean",       {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarArray",         {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarObject",        {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarKey",           {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarNull",          {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarEnumMember",    {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarStruct",        {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarEvent",         {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarOperator",      {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarTypeParameter", {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarText",          {default = true, bg = "#000000", fg = "#ffffff"})
vim.api.nvim_set_hl(0, "LspsagaWinbarSeparator",     {default = true, bg = "#000000", fg = "#ffffff"})
```

Reply:
A lot less headaches

Reply:
because it will need you write too much these codes. try to  refactor kind now .

Reply:
Really?

```lua
{
	LspsagaWinbarFile = { fg = C.text },
	LspsagaWinbarModule = { fg = C.blue },
	LspsagaWinbarNamespace = { fg = C.peach },
	LspsagaWinbarPackage = { fg = C.mauve },
	LspsagaWinbarClass = { fg = C.mauve },
	LspsagaWinbarMethod = { fg = C.mauve },
	LspsagaWinbarProperty = { fg = C.sky },
	LspsagaWinbarField = { fg = C.teal },
	LspsagaWinbarConstructor = { fg = C.blue },
	LspsagaWinbarEnum = { fg = C.green },
	LspsagaWinbarInterface = { fg = C.peach },
	LspsagaWinbarFunction = { fg = C.mauve },
	LspsagaWinbarVariable = { fg = C.blue },
	LspsagaWinbarConstant = { fg = C.sky },
	LspsagaWinbarString = { fg = C.green },
	LspsagaWinbarNumber = { fg = C.green },
	LspsagaWinbarBoolean = { fg = C.peach },
	LspsagaWinbarArray = { fg = C.blue },
	LspsagaWinbarObject = { fg = C.peach },
	LspsagaWinbarKey = { fg = C.red },
	LspsagaWinbarNull = { fg = C.red },
	LspsagaWinbarEnumMember = { fg = C.green },
	LspsagaWinbarStruct = { fg = C.mauve },
	LspsagaWinbarEvent = { fg = C.mauve },
	LspsagaWinbarOperator = { fg = C.green },
	LspsagaWinbarTypeParameter = { fg = C.green },
	LspsagaWinbarTypeAlias = { fg = C.green },
	LspsagaWinbarParameter = { fg = C.blue },
	LspsagaWinbarStaticMethod = { fg = C.peach },
	LspsagaWinbarMacro = { fg = C.red },
}
```

Reply:
Seems like an absolute win for me...

Reply:
you can call this to override in colorscheme 

```lua
require('lspsaga.lspkind').init_or_reload(custom)  custom like the `ui.kind` key is icon value is string or table. 
```

Reply:
Why `string or table`?

I thought we needed `function` support

Reply:
Ok now the colors are all white by default

![image](https://user-images.githubusercontent.com/56817415/215265898-4f2168a5-d911-4de7-a4f8-b80a1157b553.png)

Reply:
because I don't add the Colorscheme event. just something still working.

Reply:
and i think you can do it in lsp_saga.lua file 
```lua

callback = function()
  require('lspasga.lspkind).init_or_relate(custom_kind())
end

Reply:
Honestly if you want to remove user config's table just store `config.ui.kind` function, these `init_or_reload` functions are very confusing

Reply:
I have to go now, just remember that I am not happy with the current state of `kind` because it attracts a lot of monkey code and whatever you do, don't make it harder for colorschemes to support your plugin just because of "lightweight stuff"

Reply:
> just remember that I am not happy 

Are you really interesting? What am I going to remember? Why should I care about your emotions? Who are you? I thought of supported design patterns. But I'm going to rollback.   I've made simple modifications. You just need to add a callback to your stuff. You've been asking me to make huge changes. Just to support a theme.

> don't make it harder for colorschemes to support your plugin just because of "lightweight stuff"

Why do I need support with your stuff? I never use your theme. Really weird. If you can't speak politely. Go back to school and study. Don't harass me